USDOT-SDC/infrastructure
# SDC-5460_SDC-5461: Instance Scheduler Auto-Shutdown Deploy Lambda w/Terraform

## Pull Request Information
| Proposed Release Date  | Repository                                                    | Version/Tag                                                                         | Release Type  | Risk Type  | Expected Downtime  |
|------------------------|---------------------------------------------------------------|-------------------------------------------------------------------------------------|---------------|------------|--------------------|
| 2022-00-00             | [infrastructure](https://github.com/USDOT-SDC/infrastructure) | [0.3.0 vs 0.4.0](https://github.com/USDOT-SDC/infrastructure/compare/0.3.0...0.4.0) | Fast Track    | Low        | None               |

## Release
- https://github.com/USDOT-SDC/infrastructure/releases/tag/0.4.0

## Release Notes
- [Release Notes](https://github.com/USDOT-SDC/infrastructure/blob/0.4.0/RELEASE-NOTES.md)

## Summary
- Add global schedule to instance scheduler

## Work Items
- https://securedatacommons.atlassian.net/browse/SDC-5460
  - https://securedatacommons.atlassian.net/browse/SDC-5461

## Release Notifications
- None

## Cost Impacts
- System cost impacts

## User Guide Updates
- [RT Guide: Chapter 7, Workstation Scheduler](https://securedatacommons.atlassian.net/wiki/spaces/DESK/pages/2459566081)

## Process, Procedure and Work Instruction
- [Process: Development](https://securedatacommons.atlassian.net/wiki/spaces/DO/pages/1332379871)
  - [Procedure: Infrastructure](https://securedatacommons.atlassian.net/wiki/spaces/DO/pages/2437218305)
    - [WI: Infrastructure, Infrastructure Deployment](https://securedatacommons.atlassian.net/wiki/spaces/DO/pages/2501017629)

## Impacted System Components
- Instance Scheduler

## Data Load/Processing Impact
- None

## Release Type/Risk Level classification justification
- Release Type: Fast Track
- Risk Level: Low (0)

### Risk Level Justification
| Risk | Level |
|:---- |:----- |
| Are there any modifications to system configuration?                           | 0  |
| Are there any additions or modifications to security settings?                 | 0  |
| Is a new workflow introduced or an existing workflow altered significantly?    | 0  |
| Is user behavior affected as a result of this change?                          | 0  |
| Is service outage required to deploy this change?                              | 0  |
| Are new system inter-dependencies introduced?                                  | 0  |
| Has this type of change been attempted before successfully? (Yes = 0, No = 1)  | 0  |

## Plans
- https://github.com/USDOT-SDC/infrastructure/tree/0.4.0/plans

### Test Results
- Check that the following have deployed correctly
  - Lambda: instance-scheduler
- Using [sample_global_schedule.yaml](terraform\instance-scheduler\lambdas\instance-scheduler\sample_global_schedule.yaml), setup the global schedule parameter store
- Observe that instances are started and stopped by the instance-scheduler

## Code Review Checklist
- [x] **Avoid Hardcoding**  
Any string literal values that are likely to change should not be hardcoded and instead put into a config file or something like the AWS Systems Manager Parameter Store. 
This would obviously include any secrets but would also include things URLs, user names, target services, etc. For example, if posting to an SNS topic, we might consider 
externalizing the topic name to make it configurable. For cases where this seems like overkill, a good compromise is hardcoding a default value. In other words, attempt 
to read from an external configuration (e.g. Parameter Store) but use the default value if not found. This does not require much more code but gives you an option to 
change configuration later if needed.

We must avoid copying-and-pasting code in multiple places. If we have utility functions that would be useful they should be put in a common library for reuse. 
For Lambda functions, this may mean creating a Lambda layer for utility code that can be reused across multiple Lambdas. For regular Python code, this might involve 
creating a custom Python library. Even if this is the first time writing a piece of code, if it can be made generic and put into a common library/package, we should 
strongly consider doing this. 

- [x] **Self-Documenting Code**  
Code should include comments for clarity when appropriate but should also be self-documenting to the extent possible. This means using meaningful function and variable 
names to help indicate when the function or variable is being used for. Comments above a function that explain the inputs and outputs are useful but not necessary for 
code that has a good function name and can be understood by a junior developer.

Are secrets being utilized in the code? If so, are they handled properly and not logged to output or exposed. If necessary, is appropriate authentication in place for the code 
to be invoked? Is input data from non-trusted sources appropriately validated and/or sanitized?  
Check for the following:
  - Passwords
  - Private Tokens/Keys
  - Account Numbers
  - Usernames
  - Email Addresses

## Users to Notify
- None

## Security Review Needed (DOT)
- No
